### ISSUE TITLE
Emit bundler-friendly URL locators

### ISSUE THREAD

**Issue by hardfist** [emscripten-core/emscripten#13571]
according to  https://twitter.com/RReverser/status/1365342246653018115 ,wasm-bindgen already support emit code import wasm using import, which is very helpful for bundler, I wonder whether emscripten could support the same feature?

**Comment by RReverser** [comment]
To elaborate: in `EXPORT_ES6` mode, we should be able to use `new URL('...', import.meta.url)` pattern as the default Wasm location.

This pattern is increasingly commonly recognised by tooling, including major bundlers, and would allow to bundle Emscripten-generated JS more effortlessly.

It's, of course, also a valid runtime expression to get a URL of the Wasm file relatively to the module, so it works directly in browsers too.

**Comment by RReverser** [comment]
I tried to take a stab at this, but ran into a blocker on the Closure side :( https://github.com/google/closure-compiler/issues/3707

**Comment by RReverser** [comment]
No, wait, I think it's on Terser side...

**Comment by RReverser** [comment]
Ah it's both.

**Comment by kripken** [comment]
Perhaps this could be implemented in `-O0/-O1` at least, which don't run terser or closure? Then we can expand later when those projects fix things.

**Comment by RReverser** [comment]
@kripken I actually worked around that issue, but then ran into `scriptDirectory` issues on `EXPORT_ES6` + Node.js from https://github.com/emscripten-core/emscripten/issues/11792, then tried to fix those but got very confused around all the `scriptDirectory` handling and nested `#ifdef`s (there seems to be quite a bit of duplication) and that's where I'm at now 😅

Maybe if someone wants to tackle https://github.com/emscripten-core/emscripten/issues/11792, I'm happy to do just this one on top.

**Comment by curiousdannii** [comment]
Yes, unfortunately despite being there for years, EXPORT_ES6 is still fundamentally broken. :(

**Comment by matthewp** [comment]
I worked around the minifier support bug myself by instead using `IMPORT_META_URL` and then replacing that with `import.meta.url` in a post-build transform. Not sure if a hacky solution like that is worthy of being in the compiler.

**Comment by RReverser** [comment]
@matthewp Yeah I did something similar in Emscripten patch too; the Node issue / scriptDirectory still needs to be fixed though.

**Comment by matthewp** [comment]
I haven't ran into the `scriptDirectory` issue, probably because I'm using a custom `locateFile`. the worst issue I ran into is that in Node it depends on synchronous requires that can't just be replaced with dynamic import. So I wound up prepending static imports for fs/path, which meant having separate modules for Node and browser.

**Comment by RReverser** [comment]
@matthewp Yeah that issue was also mentioned in #11792.

**Comment by matthewp** [comment]
@RReverser Ah yeah, I now see what you mean about the nested `#ifs` being confusing. I wonder if that can be replaced with JS ifs and then let closure worry about DCE-ing unused conditions. Probably for another issue. I'll follow up in that other issue.

**Comment by hardfist** [comment]
> Perhaps this could be implemented in `-O0/-O1` at least, which don't run terser or closure? Then we can expand later when those projects fix things.

Yes, We use esbuild to bundle&minify wasm in production, which does not rely on terser or google closure compiler, I think wasm bundle strategy should be left to user land(so many strategies could be token),and do not think bound with terser or Closure is an good idea.

**Comment by RReverser** [comment]
@hardfist Not really; running Closure on practice is still useful regardless of the final minifier, because Closure can do a lot better minification using Emscripten-emitted annotations, whereas any regular minifier would just ignore those comments and would have less leeway in optimizing code and removing unused pieces.

**Comment by hardfist** [comment]
any one working on this ? It's really annoying to deploy an universal wasm library without this feature. and we're ok with the unminify version of glue js.

**Comment by kripken** [comment]
It looks like closure fixed this meanwhile. Perhaps terser has too? If so maybe we can update.

If terser is a blocker here, we could consider adding an option to not run it even when optimizing. The results would be less optimal (e.g. metadce would not run), but it would unblock people?

**Comment by RReverser** [comment]
Yes, there is more than Closure though - I've actually made quite a bit of progress on this (again) today and can make a PR next week that will cover most common scenarios.

**Comment by RReverser** [comment]
Sent a PR at https://github.com/emscripten-core/emscripten/pull/14135.
